### ISSUE TITLE
Accessing nested keys in JSON type variable

### ISSUE THREAD

**Issue by kellyjonbrazil** [lmorg/murex#430]
Hi there!

I'm the author of `jc` and I'm creating some [documentation](https://github.com/kellyjonbrazil/jc/wiki/Using-jc-With-Different-Shells) around integrating `jc` with other shells, like Murex. As with other shell integrations I've documented, I'd like to show a use case of assigning `jc`'s JSON output to a variable, filtering the variable, and printing it.

For example, in PowerShell:
```
PS C:\> $myvar = dig www.google.com | jc --dig | ConvertFrom-Json
PS C:\> Write-Output $myvar[0].answer[0].data
64.233.185.104
```
I think I can do something similar in Murex like this:
```
~ » jc dig www.google.com -> set json myvar
~ » out $myvar                                                                                                                       
[{"id":50161,"opcode":"QUERY","status":"NOERROR","flags":["qr","rd","ra"],"query_num":1,"answer_num":6,"authority_num":0,"additional_num":1,"opt_pseudosection":{"edns":{"version":0,"flags":[],"udp":4096}},"question":{"name":"www.google.com.","class":"IN","type":"A"},"answer":[{"name":"www.google.com.","class":"IN","type":"A","ttl":225,"data":"142.251.15.106"},{"name":"www.google.com.","class":"IN","type":"A","ttl":225,"data":"142.251.15.104"},{"name":"www.google.com.","class":"IN","type":"A","ttl":225,"data":"142.251.15.99"},{"name":"www.google.com.","class":"IN","type":"A","ttl":225,"data":"142.251.15.103"},{"name":"www.google.com.","class":"IN","type":"A","ttl":225,"data":"142.251.15.105"},{"name":"www.google.com.","class":"IN","type":"A","ttl":225,"data":"142.251.15.147"}],"query_time":86,"server":"2600:1700:bab0:d40::1#53(2600:1700:bab0:d40::1)","when":"Sun May 22 10:34:13 PDT 2022","rcvd":139,"when_epoch":1653240853,"when_epoch_utc":null}]
```
So far, so good. Now I would like to filter the variable down to a specific, nested key. It seems I can user `$myvar[0]` to get the first Object in the list, but I don't see how I can go deeper than that with the bracket syntax.:
```
# expected output: 50161
~ » out $myvar[0]["id"]                                                                                                              
{"additional_num":1,"answer":[{"class":"IN","data":"142.251.15.106","name":"www.google.com.","ttl":225,"type":"A"},{"class":"IN","data":"142.251.15.104","name":"www.google.com.","ttl":225,"type":"A"},{"class":"IN","data":"142.251.15.99","name":"www.google.com.","ttl":225,"type":"A"},{"class":"IN","data":"142.251.15.103","name":"www.google.com.","ttl":225,"type":"A"},{"class":"IN","data":"142.251.15.105","name":"www.google.com.","ttl":225,"type":"A"},{"class":"IN","data":"142.251.15.147","name":"www.google.com.","ttl":225,"type":"A"}],"answer_num":6,"authority_num":0,"flags":["qr","rd","ra"],"id":50161,"opcode":"QUERY","opt_pseudosection":{"edns":{"flags":[],"udp":4096,"version":0}},"query_num":1,"query_time":86,"question":{"class":"IN","name":"www.google.com.","type":"A"},"rcvd":139,"server":"2600:1700:bab0:d40::1#53(2600:1700:bab0:d40::1)","status":"NOERROR","when":"Sun May 22 10:34:13 PDT 2022","when_epoch":1653240853,"when_epoch_utc":null}[id]
```
As you can see, it puts `[id]` at the end instead of pulling out that key's value. Is there a way to reference nested JSON objects in native Murex syntax? In my example I'd be interested in printing `$myvar[0]['answer'][0]['data']`

Thanks,

p.s. I would also like to add this information to my blog post on [Working With JSON in Various Shells](https://blog.kellybrazil.com/2022/05/19/working-with-json-in-various-shells/).

**Comment by lmorg** [comment]
Hi @kellyjonbrazil 

There's a couple of ways this can be achieved, either by piping the index function:
```
jc dig www.google.com -> cast json -> [0] -> [id]
```
or using the double bracket "element" function which supports a full path:
```
jc dig www.google.com -> cast json -> [[/0/id]]
```

Both work as variables too `$myvar[0]->[id]` and `$myvar[[/0/id]]` (in fact the `[]` and `[[]]` is just syntactic sugar to do the same as the longer example above).

The path in "element" needs to be prefixed with a separator character. In this instance I've used `/` but you could use any ASCII character. eg `[[ .0.id ]]`

The benefit of "index" (`[]`) over `element` (`[[]]`) is that with index you can return multiple values, albeit only in the top level. Whereas `element` can reach deeper inside but for only one value. eg 

```
» jc dig www.google.com -> cast json -> [0] -> [answer question server]                                                                                                                                       
[
    [
        {
            "class": "IN",
            "data": "142.250.187.228",
            "name": "www.google.com.",
            "ttl": 194,
            "type": "A"
        }
    ],
    {
        "class": "IN",
        "name": "www.google.com.",
        "type": "A"
    },
    "192.168.1.5#53(192.168.1.5) (UDP)"
]
```

Index: https://murex.rocks/docs/commands/index.html
Element: https://murex.rocks/docs/commands/element.html

**Comment by lmorg** [comment]
@kellyjonbrazil 

Had a read through your blog. Nice work there :+1:

**Comment by kellyjonbrazil** [comment]
Running into an issue trying to test this syntax:
```
~ » jc dig www.google.com -> cast json -> set myvar                                                                                  
~ » out $myvar                                                                                                                       
[{"id":57579,"opcode":"QUERY","status":"NOERROR","flags":["qr","rd","ra"],"query_num":1,"answer_num":1,"authority_num":0,"additional_num":1,"opt_pseudosection":{"edns":{"version":0,"flags":[],"udp":4096}},"question":{"name":"www.google.com.","class":"IN","type":"A"},"answer":[{"name":"www.google.com.","class":"IN","type":"A","ttl":180,"data":"172.217.165.36"}],"query_time":41,"server":"2600:1700:bab0:d40::1#53(2600:1700:bab0:d40::1)","when":"Mon May 23 07:55:44 PDT 2022","rcvd":59,"when_epoch":1653317744,"when_epoch_utc":null}]
~ » out $myvar[[.0.id]]                                                                                                              
Error in `[[` ( 2,10): missing closing brackets, ` ]]`
]
~ »                   
```
I'm running Murex v2.7.7500 BETA via homebrew, so not sure if that might be the issue. Or my coffee just hasn't kicked in yet! :)

**Comment by kellyjonbrazil** [comment]
Thanks - yes, I just want to show a complete example of receiving JSON string data, casting and setting, and using the variable in a command. I went ahead and added Murex to the [wiki page](https://github.com/kellyjonbrazil/jc/wiki/Using-jc-With-Different-Shells#murex) and the [blog post](https://blog.kellybrazil.com/2022/05/19/working-with-json-in-various-shells/).  Thanks again for your help!

**Comment by lmorg** [comment]
Looks good!

Fixed support for $MyVar[[/element]] in f07fa7c2dd2d5a3ee5ec9372eb88280c6ce7ae27

That's currently being built so should the website should have the latest versions in about half an hour. Homebrew is usually a day or two behind
